gaps/gap-8_basic_auth_runtime/gap-8_basic_auth_runtime.md

```diff
@@ -1,41 +1,190 @@
+
+[log-specific var reference](https://nginx.org/en/docs/http/ngx_http_log_module.html#log_format)
+
+### Runtime implementation
+
+Base the implementation on [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk)
+- use [ya-runtime-basic-auth](https://github.com/golemfactory/yagna-service-erigon/tree/master/ya-runtime-basic-auth) as a starting point
+    - refer to [this comment](https://github.com/golemfactory/yagna-service-erigon/pull/47#issuecomment-983924928) for general guidance
+    - (htpasswd) port the `htpasswd.sh` script to Rust, e.g. using the `htpasswd-verify` crate, an implementation of the `apr1` digest and a `.htpasswd` formatter
+
+#### Environment
+
+Use a custom implementation of an `Env` trait for dynamically setting the runtime name and thus, the automatic choosing of the configuration file during startup.
+
+An example can be found [here](https://github.com/golemfactory/ya-runtime-sdk/blob/mf/dynamic-rt-name/examples/environment/src/main.rs).
+
+#### Command line arguments
+
+Command line arguments should accept a runtime name parameter.
+
+An example can be found [here](https://github.com/golemfactory/ya-runtime-sdk/blob/mf/dynamic-rt-name/examples/environment/src/main.rs).
+
+#### Configuration
+
+The configuration `struct` / `enum` should be de-serializable from all variants of configuration files (i.e., all service types) and contain the following information:
+
+- service binary location
+- nginx access log location
+- nginx `.htpasswd` file location
+- HTTP URIs reserved for the service (regex?)
+
+#### Background jobs
+
+1. Nginx log parser
+   
+   Monitors the nginx access log file and counts the number of requests made by the current user.
+
+2. Daemon counter updater
+   
+   Periodically updates the custom counter value (total requests made by the user), extracted from the nginx access log file.
+   An example of a custom counter can be found [here](https://github.com/golemfactory/ya-test-runtime-counters/blob/main/src/main.rs).
+
+#### `Runtime` trait implementation
+
+1. `deploy`
+
+- **in background**, start an instance of `erigolem` / `lighthouse` (if not running)
+  and wait until the application is ready to accept commands.
+
+2. `start`
+
+- none
+
+3. `run`
+
+    - `credentials <username> <password>`
+
+      Sets the credentials that will be used in the current activity. If a username exists,
+      returns an error (with a proper message outputted to `stderr`); the runtime is terminated.
+
+    - `counter <name> <username> <password>` (optional)
+
+      Reads the specific counter value for a given user. The foreseen counter
+      names will be set to the requested URIs (locations) configured in nginx.
+
+4. `transfer`
+
+    - none
+
+5. `offer-template`
+
+- if the server supports HTTPS, include a `https` runtime capability in an offer
+- (optional) in case of self-signed certificates, add a property containing
+  the certificate (name TBD)
```

Comment:
The services do not speak the property language - we don't know what can be published, how it should be named or what the service actually does. We also cannot simply forward service configuration files since they may contain sensitive data

Reply:
On a second thought, we could add a property section to configuration file(s). There are some problems though:
- they need to be created by hand; property names and their value semantics would have to be somehow known to requestors
- if values change, they need to be updated by hand
- there are currently no uses for this feature

but I can imagine this could become useful in the future

Reply:
Why we can't do like in `ya-runtime-vm` where runtime is returning Offer?